src/cargo/core/global_cache_tracker.rs

```diff
@@ -0,0 +1,1754 @@
+//! file.
+//!
+//! While cargo is running, when it detects a use of a cache file, it adds a
+//! timestamp to [`DeferredGlobalLastUse`]. This batches up a set of changes
+//! that are then flushed to the database all at once (via
+//! [`DeferredGlobalLastUse::save`]). Ideally saving would only be done once
+//! for performance reasons, but that is not really possible due to the way
+//! cargo works, since there are different ways cargo can be used (like `cargo
+//! generate-lockfile`, `cargo fetch`, and `cargo build` are all very
+//! different ways the code is used).
+//!
+//! All of the database interaction is done through the [`GlobalCacheTracker`]
+//! type.
+//!
+//! There is a single global [`GlobalCacheTracker`] and
+//! [`DeferredGlobalLastUse`] stored in [`Config`].
+//!
+//! ## Automatic gc
+//!
+//! Some commands (primarily the build commands) will trigger an automatic
+//! deletion of files that haven't been used in a while. The interface for
+//! this is in the [`crate::core::gc`] module. The database tracks the last
+//! time an automatic gc was performed so that it is only done once per day
+//! for performance reasons.
+//!
+//! ## Manual gc
+//!
+//! The user can perform a manual garbage collection with the `cargo clean`
+//! command. That command has a variety of options to specify what to delete.
+//! Manual gc supports deleting based on age or size or both.
+//!
+//! ## Locking
+//!
+//! Usage of the database requires that the package cache is locked to prevent
+//! concurrent access. Although sqlite has built-in locking support, we want
+//! to use cargo's locking so that the "Blocking" message gets displayed, and
+//! so that locks can block indefinitely for long-running build commands.
+//! [`rusqlite`] has a default timeout of 5 seconds, though that is
+//! configurable.
+//!
+//! When garbage collection is being performed, the package cache lock must be
+//! in [`CacheLockMode::MutateExclusive`] to ensure no other cargo process is
+//! running. See [`crate::util::cache_lock`] for more detail on locking.
+//!
+//! ## Compatibility
+//!
+//! The database must retain both forwards and backwards compatibility between
+//! different versions of cargo. For the most part, this shouldn't be too
+//! difficult to maintain. Generally sqlite doesn't change on-disk formats
+//! between versions (the introduction of WAL is one of the few examples where
+//! version 3 had a format change, but we wouldn't use it anyway since it has
+//! shared-memory requirements cargo can't depend on due to things like
+//! network mounts).
+//!
+//! Schema changes must be managed through [`migrations`] by adding new
+//! entries that make a change to the database. Changes must not break older
+//! versions of cargo. Generally, adding columns should be fine (either with a
+//! default value, or NULL). Adding tables should also be fine. Just don't do
+//! destructive things like removing a column, or changing the semantics of an
+//! existing column.
```

Comment:
How are we expected to deal with adding new columns because we intend to use them but old cargo will ignore those columns?

Reply:
For new columns, newer cargos will need to assume that the column might be filled in with the default (which can be NULL if that is how it is defined).  For new tables, it will just need to assume that the data is not managed by older cargos.